src/main/java/io/confluent/kafkarest/v2/KafkaConsumerState.java

```diff
@@ -0,0 +1,451 @@
+/**
+ * Copyright 2017 Confluent Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ **/
+package io.confluent.kafkarest.v2;
+
+import io.confluent.kafkarest.ConsumerInstanceId;
+import io.confluent.kafkarest.ConsumerRecordAndSize;
+import io.confluent.kafkarest.KafkaRestConfig;
+import io.confluent.kafkarest.entities.ConsumerAssignmentRequest;
+import io.confluent.kafkarest.entities.ConsumerCommittedRequest;
+import io.confluent.kafkarest.entities.ConsumerCommittedResponse;
+import io.confluent.kafkarest.entities.ConsumerOffsetCommitRequest;
+import io.confluent.kafkarest.entities.ConsumerSeekToOffsetRequest;
+import io.confluent.kafkarest.entities.ConsumerSeekToRequest;
+import io.confluent.kafkarest.entities.ConsumerSubscriptionRecord;
+import io.confluent.kafkarest.entities.TopicPartitionOffset;
+import io.confluent.kafkarest.entities.TopicPartitionOffsetMetadata;
+import kafka.serializer.Decoder;
+import org.apache.kafka.clients.consumer.Consumer;
+import org.apache.kafka.clients.consumer.ConsumerRebalanceListener;
+import org.apache.kafka.clients.consumer.ConsumerRecord;
+import org.apache.kafka.clients.consumer.ConsumerRecords;
+import org.apache.kafka.clients.consumer.OffsetAndMetadata;
+import org.apache.kafka.common.TopicPartition;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.Vector;
+import java.util.concurrent.locks.ReadWriteLock;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+import java.util.regex.Pattern;
+
+/**
+ * Tracks all the state for a consumer. This class is abstract in order to support multiple
+ * serialization formats. Implementations must provide decoders and a method to convert Kafka
+ * MessageAndMetadata<K,V> values to ConsumerRecords that can be returned to the client (including
+ * translation if the decoded Kafka consumer type and ConsumerRecord types differ).
+ */
+public abstract class KafkaConsumerState<KafkaK, KafkaV, ClientK, ClientV>
+    implements Comparable<KafkaConsumerState> {
+
+  private KafkaRestConfig config;
+  private ConsumerInstanceId instanceId;
+  private Consumer consumer;
+  private ConsumerRecords<ClientK, ClientV> consumerRecords = null;
```

Comment:
I don't see any reason why a Java queue can't be used instead of a list and building peek() and next() functions on top a list